File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureActivityUnitTest.java
package com.braintreepayments.api.threedsecure;

import static android.app.Activity.RESULT_OK;
import static com.braintreepayments.api.threedsecure.ThreeDSecureActivity.RESULT_COULD_NOT_START_CARDINAL;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.content.Intent;
import android.os.Bundle;

import com.braintreepayments.api.testutils.Fixtures;
import com.braintreepayments.api.core.BraintreeException;
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalActionCode;
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalChallengeObserver;
import com.cardinalcommerce.cardinalmobilesdk.models.ValidateResponse;

import org.json.JSONException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class ThreeDSecureActivityUnitTest {

    @Test
    public void onCreate_withExtras_invokesCardinalWithLookupData()
            throws JSONException, BraintreeException {
        ThreeDSecureParams threeDSecureParams =
                ThreeDSecureParams.fromJson(Fixtures.THREE_D_SECURE_LOOKUP_RESPONSE);

        Bundle extras = new Bundle();
        extras.putParcelable(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT, threeDSecureParams);

        Intent intent = new Intent();
        intent.putExtras(extras);

        ThreeDSecureActivity sut = new ThreeDSecureActivity();
        sut.setIntent(intent);

        CardinalClient cardinalClient = mock(CardinalClient.class);
        sut.launchCardinalAuthChallenge(cardinalClient);

        ArgumentCaptor<ThreeDSecureParams> captor =
                ArgumentCaptor.forClass(ThreeDSecureParams.class);
        verify(cardinalClient).lookupChallenge(captor.capture(), any());

        ThreeDSecureParams actualResult = captor.getValue();
        ThreeDSecureLookup actualLookup = actualResult.getLookup();
        assertEquals("sample-transaction-id", actualLookup.getTransactionId());
        assertEquals("sample-pareq", actualLookup.getPareq());
    }

    @Test
    public void onCreate_withExtrasAndCardinalError_finishesWithError()
            throws JSONException, BraintreeException {
        ThreeDSecureParams threeDSecureParams =
                ThreeDSecureParams.fromJson(Fixtures.THREE_D_SECURE_LOOKUP_RESPONSE);

        Bundle extras = new Bundle();
        extras.putParcelable(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT, threeDSecureParams);

        Intent intent = new Intent();
        intent.putExtras(extras);

        ThreeDSecureActivity sut = spy(new ThreeDSecureActivity());
        sut.setIntent(intent);

        BraintreeException cardinalError = new BraintreeException("fake cardinal error");

        CardinalClient cardinalClient = mock(CardinalClient.class);
        doThrow(cardinalError).when(cardinalClient)
                .lookupChallenge(any(ThreeDSecureParams.class), any());

        sut.launchCardinalAuthChallenge(cardinalClient);
        verify(sut).finish();

        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);
        verify(sut).setResult(eq(RESULT_COULD_NOT_START_CARDINAL), captor.capture());

        Intent intentForResult = captor.getValue();
        assertEquals("fake cardinal error",
                intentForResult.getStringExtra(ThreeDSecureActivity.EXTRA_ERROR_MESSAGE));
    }

    @Test
    public void onCreate_withoutExtras_finishesWithError() throws BraintreeException {
        Intent intent = new Intent();
        ThreeDSecureActivity sut = spy(new ThreeDSecureActivity());
        sut.setIntent(intent);

        CardinalClient cardinalClient = mock(CardinalClient.class);
        sut.launchCardinalAuthChallenge(cardinalClient);

        verify(cardinalClient, never()).lookupChallenge(any(ThreeDSecureParams.class),
                any(CardinalChallengeObserver.class));
        verify(sut).finish();

        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);
        verify(sut).setResult(eq(RESULT_COULD_NOT_START_CARDINAL), captor.capture());

        Intent intentForResult = captor.getValue();
        assertEquals("Unable to launch 3DS authentication.",
                intentForResult.getStringExtra(ThreeDSecureActivity.EXTRA_ERROR_MESSAGE));
    }

    @Test
    public void handleValidated_returnsValidationResults() throws JSONException {
        ThreeDSecureParams threeDSecureParams =
                ThreeDSecureParams.fromJson(Fixtures.THREE_D_SECURE_LOOKUP_RESPONSE);

        Bundle extras = new Bundle();
        extras.putParcelable(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT, threeDSecureParams);
        Intent intent = new Intent();
        intent.putExtras(extras);
        ThreeDSecureActivity sut = spy(new ThreeDSecureActivity());
        sut.setIntent(intent);

        ValidateResponse cardinalValidateResponse = mock(ValidateResponse.class);
        when(cardinalValidateResponse.getActionCode()).thenReturn(CardinalActionCode.SUCCESS);
        CardinalClient cardinalClient = mock(CardinalClient.class);
        sut.handleValidated(cardinalClient, cardinalValidateResponse, "jwt");
        verify(sut).finish();

        ArgumentCaptor<Intent> captor = ArgumentCaptor.forClass(Intent.class);
        verify(sut).setResult(eq(RESULT_OK), captor.capture());
        verify(cardinalClient).cleanup();

        Intent intentForResult = captor.getValue();
        ValidateResponse activityResult = (ValidateResponse) (intentForResult.getSerializableExtra(ThreeDSecureActivity.EXTRA_VALIDATION_RESPONSE));
        assertEquals("jwt", intentForResult.getStringExtra(ThreeDSecureActivity.EXTRA_JWT));
        assertEquals(threeDSecureParams, intentForResult.getParcelableExtra(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT));
        assertNotNull(activityResult);
        assertEquals("SUCCESS", activityResult.getActionCode().getString());
    }
}

File: ThreeDSecure/src/test/java/com/braintreepayments/api/threedsecure/ThreeDSecureActivityUnitTest.kt
package com.braintreepayments.api.threedsecure

import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.Bundle
import com.braintreepayments.api.core.BraintreeException
import com.braintreepayments.api.testutils.Fixtures
import com.braintreepayments.api.threedsecure.ThreeDSecureActivity.Companion.RESULT_COULD_NOT_START_CARDINAL
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalActionCode
import com.cardinalcommerce.cardinalmobilesdk.models.CardinalChallengeObserver
import com.cardinalcommerce.cardinalmobilesdk.models.ValidateResponse
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.spyk
import io.mockk.verify
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull

@RunWith(RobolectricTestRunner::class)
class ThreeDSecureActivityUnitTest {

    @Test
    fun `onCreate with extras invokes cardinal with lookup data`() {
        val threeDSecureParams = ThreeDSecureParams.fromJson(Fixtures.THREE_D_SECURE_LOOKUP_RESPONSE)

        val extras = Bundle()
        extras.putParcelable(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT, threeDSecureParams)

        val intent = Intent()
        intent.putExtras(extras)

        val sut = ThreeDSecureActivity()
        sut.intent = intent

        val cardinalClient = mockk<CardinalClient>(relaxed = true)
        sut.launchCardinalAuthChallenge(cardinalClient)

        val paramsSlot = slot<ThreeDSecureParams>()
        verify { cardinalClient.continueLookup(capture(paramsSlot), any()) }

        val actualResult = paramsSlot.captured
        val actualLookup = actualResult.lookup
        assertEquals("sample-transaction-id", actualLookup?.transactionId)
        assertEquals("sample-pareq", actualLookup?.pareq)
    }

    @Test
    fun `onCreate with extras and cardinal error finishes with error`() {
        val threeDSecureParams = ThreeDSecureParams.fromJson(Fixtures.THREE_D_SECURE_LOOKUP_RESPONSE)

        val extras = Bundle()
        extras.putParcelable(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT, threeDSecureParams)

        val intent = Intent()
        intent.putExtras(extras)

        val sut = spyk<ThreeDSecureActivity>()
        sut.intent = intent

        val cardinalError = BraintreeException("fake cardinal error")
        val cardinalClient = mockk<CardinalClient>(relaxed = true)
        every { cardinalClient.continueLookup(any<ThreeDSecureParams>(), any()) } throws cardinalError

        sut.launchCardinalAuthChallenge(cardinalClient)
        verify { sut.finish() }

        val slot = slot<Intent>()
        verify { sut.setResult(RESULT_COULD_NOT_START_CARDINAL, capture(slot)) }

        val intentForResult = slot.captured
        assertEquals("fake cardinal error", intentForResult.getStringExtra(ThreeDSecureActivity.EXTRA_ERROR_MESSAGE))
    }

    @Test
    fun `onCreate without extras finishes with error`() {
        val intent = Intent()
        val sut = spyk<ThreeDSecureActivity>()
        sut.intent = intent

        val cardinalClient = mockk<CardinalClient>(relaxed = true)
        sut.launchCardinalAuthChallenge(cardinalClient)

        verify(exactly = 0) {
            cardinalClient.continueLookup(any<ThreeDSecureParams>(), any<CardinalChallengeObserver>())
        }
        verify { sut.finish() }

        val slot = slot<Intent>()
        verify { sut.setResult(RESULT_COULD_NOT_START_CARDINAL, capture(slot)) }

        val intentForResult = slot.captured
        assertEquals(
            "Unable to launch 3DS authentication.",
            intentForResult.getStringExtra(ThreeDSecureActivity.EXTRA_ERROR_MESSAGE)
        )
    }

    @Test
    fun `handleValidated returns validation results`() {
        val threeDSecureParams = ThreeDSecureParams.fromJson(Fixtures.THREE_D_SECURE_LOOKUP_RESPONSE)

        val extras = Bundle()
        extras.putParcelable(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT, threeDSecureParams)
        val intent = Intent()
        intent.putExtras(extras)
        val sut = spyk<ThreeDSecureActivity>()
        sut.intent = intent

        val cardinalValidateResponse = mockk<ValidateResponse>(relaxed = true)
        every { cardinalValidateResponse.actionCode } returns CardinalActionCode.SUCCESS
        val cardinalClient = mockk<CardinalClient>(relaxed = true)
        sut.handleValidated(cardinalClient, cardinalValidateResponse, "jwt")
        verify { sut.finish() }

        val slot = slot<Intent>()
        verify { sut.setResult(RESULT_OK, capture(slot)) }
        verify { cardinalClient.cleanup() }

        val intentForResult = slot.captured
        val activityResult =
            intentForResult.getSerializableExtra(ThreeDSecureActivity.EXTRA_VALIDATION_RESPONSE) as ValidateResponse
        assertEquals("jwt", intentForResult.getStringExtra(ThreeDSecureActivity.EXTRA_JWT))
        assertEquals(
            threeDSecureParams,
            intentForResult.getParcelableExtra(ThreeDSecureActivity.EXTRA_THREE_D_SECURE_RESULT)
        )
        assertNotNull(activityResult)
        assertEquals("SUCCESS", activityResult.actionCode.string)
    }
}
